### PR TITLE
Prometheus: Misc updates

### PR DESCRIPTION
**Pull request by Baughn**
###### Things done:
- [X] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- [X] Built on platform(s): NixOS
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
###### More

cc @benley 

This bumps Prometheus past the recent API change to Alertmanager. 0.17.0 is still a release candidate, but it's been solid in my experience, and the changes are relevant to my plans for integrating it with NixOS. (I don't want to write alertmanager integration twice.)

### CODE REVIEW

File: pkgs/top-level/go-packages.nix [hunk 1]
```diff
@@ -2671,6 +2665,9 @@ let
       pushover
     ];
 
+    # Tests exist, but seem to clash with the firewall.
+    doCheck = false;
+
     buildFlagsArray = ''
       -ldflags=
           -X main.buildVersion=${rev}
```
Comment (benley) [inline pkgs/top-level/go-packages.nix]: It looks like the metadata ldflags used in the alertmanager build have changed in the rewrite: https://github.com/prometheus/alertmanager/blob/master/scripts/build.sh#L31

Those are just used on the status page I believe, but it's nice to have some useful data in there.

File: pkgs/top-level/go-packages.nix [hunk 2]
```diff
@@ -2898,24 +2890,23 @@ let
       yaml-v2
     ];
 
+    doCheck = true;
+
+    buildFlagsArray = ''
+      -ldflags=
+          -X main.buildVersion=${rev}
```
Comment (benley) [inline pkgs/top-level/go-packages.nix]: These don't match what prometheus uses upstream - does it still show the right metadata on the status page at runtime?

(https://github.com/prometheus/prometheus/blob/0.17.0rc2/scripts/build.sh#L32)

**Comment by benley** [comment]
Thanks! This looks pretty good, just a couple of comments about ldflags.

**Comment by Baughn** [comment]
I noticed the change, then copied the necessary update the *wrong way*, overwriting the newer code. ...oops. Thanks for spotting that.

Should be good now. Example Prometheus: https://status.brage.info/status

**Comment by Baughn** [comment]
Actually, don't merge this yet. I found a bit of a problem: http://sprunge.us/KiTU

**Comment by Baughn** [comment]
Fixed.

File: pkgs/top-level/go-packages.nix [hunk 3]
```diff
@@ -2671,13 +2665,21 @@ let
       pushover
     ];
 
-    buildFlagsArray = ''
+    # Tests exist, but seem to clash with the firewall.
+    doCheck = false;
+
+    preBuild = ''
+      export GO15VENDOREXPERIMENT=1
```
Comment (benley) [inline pkgs/top-level/go-packages.nix]: if you're enabling this, I think it means all the golang buildInputs are ignored in favor of the vendor/ directory in the prometheus tree.  I'm not sure what the current policy (if there is one) on that - but it's probably at least encouraged to use the nixpkgs version of library dependencies if possible.

File: pkgs/top-level/go-packages.nix [hunk 4]
```diff
@@ -2651,17 +2651,11 @@ let
...
-      sha256 = "0g656rzal7m284mihqdrw23vhs7yr65ax19nvi70jl51wdallv15";
-    };
+  prometheus.alertmanager = buildFromGitHub rec {
+    rev = "0.1.0";
+    owner = "prometheus";
+    repo = "alertmanager";
+    sha256 = "1ya465bns6cj2lqbipmfm13wz8kxii5h9mm7lc0ba1xv26xx5zs7";
 
     buildInputs = [
       fsnotify.v0
```
Comment (benley) [inline pkgs/top-level/go-packages.nix]: I think fsnotify is no longer used in alertmanger

File: pkgs/top-level/go-packages.nix (hunk 3 above)
Comment (Baughn) [inline pkgs/top-level/go-packages.nix]: We're building binaries, not libraries, and they're statically linked. My own inclination is that the Prometheus devs know what they're doing, so let's use the version of the dependencies that they developed and tested with.

Comment (benley) [inline pkgs/top-level/go-packages.nix]: sounds sane to me.

**Comment by benley** [comment]
It looks like prometheus 0.17.0 has just been released.  Care to update the patch for the final release before I merge it?

**Comment by Baughn** [comment]
Sounds good to me. Aaand... done.

I also disabled all the dependencies for prometheus, because... well, it doesn't use them if everything is vendored.

**Pull request merged**